src/graphapi/matmul.cpp

```diff
@@ -104,5 +105,310 @@ void BackendMatmulDescriptor::getAttribute(miopenBackendAttributeName_t attribut
     }
 }
 
+OperationMatmulBuilder& OperationMatmulBuilder::setA(Tensor* A)
+{
+    mOperationMatmul.mA = checkPtr(A);
+    if(mOperationMatmul.mA->getDimensions().size() < 2)
+        MIOPEN_THROW(miopenStatusBadParm);
+    aSet = true;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setB(Tensor* B)
+{
+    mOperationMatmul.mB = checkPtr(B);
+    if(mOperationMatmul.mB->getDimensions().size() < 2)
+        MIOPEN_THROW(miopenStatusBadParm);
+    bSet = true;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setC(Tensor* C)
+{
+    mOperationMatmul.mC = checkPtr(C);
+    if(mOperationMatmul.mC->getDimensions().size() < 2)
+        MIOPEN_THROW(miopenStatusBadParm);
+    cSet = true;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setBatchCount(int64_t count)
+{
+    mOperationMatmul.mBatchCount = count;
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setGemmMOverride(Tensor* overrideTensor)
+{
+    mOperationMatmul.mGemmMOverride = checkPtr(overrideTensor);
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setGemmNOverride(Tensor* overrideTensor)
+{
+    mOperationMatmul.mGemmNOverride = checkPtr(overrideTensor);
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setGemmKOverride(Tensor* overrideTensor)
+{
+    mOperationMatmul.mGemmKOverride = checkPtr(overrideTensor);
+    return *this;
+};
+OperationMatmulBuilder& OperationMatmulBuilder::setMatmulDescriptor(Matmul* mMatmul)
+{
+    mOperationMatmul.mMatmul = checkPtr(mMatmul);
+    matmulSet                = true;
+    return *this;
+}
+
+OperationMatmul OperationMatmulBuilder::build()
+{
+    if(!aSet || !bSet || !cSet || !matmulSet)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    int aDimensionsCount = mOperationMatmul.mA->getDimensions().size();
+    int bDimensionsCount = mOperationMatmul.mB->getDimensions().size();
+    int cDimensionsCount = mOperationMatmul.mC->getDimensions().size();
+
+    if(aDimensionsCount != bDimensionsCount || bDimensionsCount != cDimensionsCount)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    int Am = mOperationMatmul.mA->getDimensions().end()[-2];
+    int An = mOperationMatmul.mA->getDimensions().end()[-1];
+
+    int Bn = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Bk = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    int Cm = mOperationMatmul.mB->getDimensions().end()[-2];
+    int Ck = mOperationMatmul.mB->getDimensions().end()[-1];
+
+    if(Am != Cm || An != Bn || Bk != Ck)
+        MIOPEN_THROW(miopenStatusBadParm);
+
+    // TODO: need broadcsat checks
+    //
```

Comment:
I did the broadcast checks for pointwise descriptor here [pointwise.cpp](https://github.com/ROCm/MIOpen/blob/0334279c5214329acd8f7dfe20b31182e17c6bdb/src/graphapi/pointwise.cpp#L686-L712). Could you check whether it fits your needs? If so, reuse the template and move it to one of the common headers, probably `graphapi.hpp`.

Reply:
fixed